lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp

```diff
@@ -910,16 +911,470 @@ static std::string GetClangModulesCacheProperty() {
+                  llvm::StringRef in_bufferName, unsigned in_bufferID,
+                  uint32_t in_line, uint32_t in_column,
+                  llvm::ArrayRef<swift::Diagnostic::FixIt> in_fixits)
+        : description(in_desc), kind(in_kind), bufferName(in_bufferName),
+          bufferID(in_bufferID), line(in_line), column(in_column),
+          fixits(in_fixits) {}
+    RawDiagnostic(const RawDiagnostic &other)
+        : description(other.description), kind(other.kind),
+          bufferName(other.bufferName), bufferID(other.bufferID),
+          line(other.line), column(other.column), fixits(other.fixits) {}
+    bool operator==(const RawDiagnostic& other) {
+      return kind == other.kind && bufferID == other.bufferID &&
+             line == other.line && column == other.column &&
+             description == other.description;
+    }
+    bool operator!=(const RawDiagnostic &other) { return !(*this == other); }
+
+    std::string description;
+    swift::DiagnosticKind kind = swift::DiagnosticKind::Error;
+    llvm::StringRef bufferName;
+    unsigned bufferID;
+    uint32_t line;
+    uint32_t column;
+    std::vector<swift::DiagnosticInfo::FixIt> fixits;
+  };
+
+  SwiftASTContext &m_ast_context;
+  /// Stores all diagnostics coming from the Swift compiler.
+  std::vector<RawDiagnostic> m_raw_swift_diagnostics;
+  std::vector<RawDiagnostic> m_raw_clang_diagnostics;
+  /// Stores diagnostics coming from LLDB.
+  std::vector<std::unique_ptr<Diagnostic>> m_diagnostics;
+
+  unsigned m_num_swift_errors = 0;
+  unsigned m_num_clang_errors = 0;
+  bool m_colorize = false;
+
+  friend class SwiftASTContext::TransactionDiagnostics;
+};
+
+} // namespace lldb_private
+
+SwiftASTContext::TransactionDiagnostics::TransactionDiagnostics(
+    swift::DiagnosticConsumer &consumer)
+    : m_consumer(consumer),
+      m_cursor({static_cast<StoringDiagnosticConsumer *>(&m_consumer)
+                    ->m_raw_swift_diagnostics.size(),
+                static_cast<StoringDiagnosticConsumer *>(&m_consumer)
+                    ->m_raw_clang_diagnostics.size(),
+                static_cast<StoringDiagnosticConsumer *>(&m_consumer)
+                    ->m_diagnostics.size()}) {}
+
+void SwiftASTContext::TransactionDiagnostics::PrintDiagnostics(
+    DiagnosticManager &diagnostic_manager, uint32_t bufferID,
+    uint32_t first_line, uint32_t last_line) const {
+  auto consumer = static_cast<StoringDiagnosticConsumer *>(&m_consumer);
+  consumer->PrintDiagnostics(diagnostic_manager, m_cursor, bufferID, first_line,
+                             last_line);
+}
+
+bool SwiftASTContext::TransactionDiagnostics::HasErrors() const {
+  auto &consumer = *static_cast<StoringDiagnosticConsumer *>(&m_consumer);
+  size_t num_errors = 0;
+  for (size_t i = m_cursor.lldb; i < consumer.m_diagnostics.size(); ++i)
+    if (consumer.m_diagnostics[i]->GetSeverity() == eDiagnosticSeverityError)
+      num_errors++;
+
+  return num_errors > 0 ||
```

Comment:
You don't need to count the number of errors if the last two conditions are true. You could do an early return:

```
 if (consumer.m_num_swift_errors > m_cursor.m_num_swift_errors ||
          consumer.m_raw_clang_diagnostics.size() > m_cursor.clang)
  return true;
```

and compute `num_errors` after.